### PR TITLE
Fix crash in node when mixing sync/async resolvers

### LINKED ISSUE #3528: Unhandled rejection if an object has both a synchronous and asynchronous "Cannot return null for non-nullable field" error
Sorry about the title but it's a pretty specific scenario.

It looks like this error occurs because of how errors are handled for non-nullable fields. (I saw this with the `Cannot return null for non-nullable field` error, but it seems if the resolver throws an error it would be handled the same way.)

The problem appears to be here: https://github.com/graphql/graphql-js/blob/main/src/execution/execute.ts#L450
Let's say in the example below, the City object type has a resolver for `country`, which is asynchronous, and a resolver for `population`, which is synchronous, and they _both_ return null (which they shouldn't, of course). The resolver for the `country` field is run, and returns a promise. Then the resolver for the `population` field is run, and because it's synchronous, the error is thrown straight away, and the promise returned by the `country` resolver is discarded. When that promise eventually rejects, the process crashes with an `unhandledRejection` error. 

```js
// File: testcase.mjs
import { graphql, GraphQLInt, GraphQLList, GraphQLNonNull, GraphQLObjectType, GraphQLSchema, GraphQLString } from "graphql";

process.on("unhandledRejection", reason => {
    console.error("Unhandled rejection!", reason);
});

const City = new GraphQLObjectType({
    name: "City",
    fields: {
        name: { type: new GraphQLNonNull(GraphQLString) },
        country: { type: new GraphQLNonNull(GraphQLString), resolve: () => Promise.resolve(null) },
        population: { type: new GraphQLNonNull(GraphQLInt), resolve: () => null },
    }
});

const Query = new GraphQLObjectType({
    name: "Query",
    fields: {
        cities: {
            type: new GraphQLNonNull(new GraphQLList(new GraphQLNonNull(City))),
            resolve() {
                return [{ name: "London", country: "GB", population: 8_982_000 }];
            }
        }
    }
});

const schema = new GraphQLSchema({
    types: [City],
    query: Query,
});

try {
    graphql(schema, `
        query {
            cities { name country population }
        }
    `).then(result => {
        if (result.errors)
            console.error("Error returned in result.errors[]:", result.errors);
        else
            console.log("Result:", result.data);
    }, error => {
        console.error("graphql() rejected:", error);
    });
} catch(err) {
    console.error("Error caught by try/catch:", err);
}
```

The output of this is:

```
Error returned in result.errors[]: [
  Error: Cannot return null for non-nullable field City.population.
      at completeValue (/home/lee/restapi/node_modules/graphql/execution/execute.js:566:13)
      at resolveField (/home/lee/restapi/node_modules/graphql/execution/execute.js:478:19)
      at executeFields (/home/lee/restapi/node_modules/graphql/execution/execute.js:292:18)
      at collectAndExecuteSubfields (/home/lee/restapi/node_modules/graphql/execution/execute.js:755:10)
      at completeObjectValue (/home/lee/restapi/node_modules/graphql/execution/execute.js:745:10)
      at completeValue (/home/lee/restapi/node_modules/graphql/execution/execute.js:597:12)
      at completeValue (/home/lee/restapi/node_modules/graphql/execution/execute.js:563:21)
      at /home/lee/restapi/node_modules/graphql/execution/execute.js:627:25
      at Array.map (<anonymous>)
      at safeArrayFrom (/home/lee/restapi/node_modules/graphql/jsutils/safeArrayFrom.js:36:23) {
    locations: [ [Object] ],
    path: [ 'cities', 0, 'population' ]
  }
]
Unhandled rejection! Error: Cannot return null for non-nullable field City.country.
    at completeValue (/home/lee/restapi/node_modules/graphql/execution/execute.js:566:13)
    at /home/lee/restapi/node_modules/graphql/execution/execute.js:475:16 {
  locations: [ { line: 3, column: 27 } ],
  path: [ 'cities', 0, 'country' ]
}
```

### PR DESCRIPTION
**Pull request by asztal**
Fixes #3528

Ensures that if `executeFields` encounters a mix of promises and thrown errors, the promises will be awaited before throwing the error.

This does technically change `executeFields` to return a rejected promise in this scenario as opposed to throwing the not-null error synchronously, but I figured if any of the resolvers are asynchronous then returning a promise will be expected anyway, and this was better than crashing the process.

Note: I had to add an `unhandledRejection` event listener myself, as it seems mocha doesn't do it.

### CODE REVIEW

File: src/execution/__tests__/executor-test.ts [hunk 1]
```diff
@@ -625,6 +625,60 @@ describe('Execute: Handles basic execution tasks', () => {
...
+            message:
+              'Cannot return null for non-nullable field Type.asyncNullError.',
+            locations: [{ line: 3, column: 11 }],
+            path: ['asyncNullError'],
+          },
+        ],
+      });
+
+      // Allow time for the asyncReject field to be rejected
+      await new Promise((resolve) => setTimeout(resolve, 50));
```
Comment (n1ru4l) [inline src/execution/__tests__/executor-test.ts:676]: Instead of some arbitrary timer `process.nextTick` or `setImmediate` could be better here.

Comment (asztal) [reply src/execution/__tests__/executor-test.ts]: Thanks, you're right, I thought I'd tried that and it didn't work for some reason, but `process.nextTick` does indeed work fine.

**Review by yaacovCR — COMMENTED** [review]
Do we need similar logic when dealing with completion of lists?

**Comment by IvanGoncharov** [comment]
@asztal  I fixed everything except CLA.
Please sign it and this PR is ready to merge

**Comment by asztal** [comment]
Thanks! My contribution counts as a corporate contribution so I need a director to approve it for me first - it should be soon hopefully 🤞

**Comment by IvanGoncharov** [comment]
Hi @asztal 👋 
Any news on CLA?

**Comment by chrskrchr** [comment]
@IvanGoncharov et al, once this fix lands in `main`, could we backport it to the `15.x` branch?  Happy to submit that PR if there aren't any objections and if @asztal doesn't want to do it themselves.

**Comment by chrskrchr** [comment]
@asztal - any word on the CLA?

**Comment by IvanGoncharov** [comment]
> @IvanGoncharov et al, once this fix lands in main, could we backport it to the 15.x branch? Happy to submit that PR if there aren't any objections and if @asztal doesn't want to do it themselves.

Sounds great 👍 
This is the exact idea behind the current dev flow.

**Comment by asztal** [comment]
> @asztal - any word on the CLA?

The CLA manager at my company is having trouble using the Corporate CLA console -- I'm not at all familiar with it, so I've asked them to check they've signed up using the correct email address, otherwise just raise a support ticket about it like the message says. Sorry for the delay.

Here's the actual error in case anyone here knows anything about it.
image

**Comment by chrskrchr** [comment]
Submitted a PR to backport this fix to `15.x`: #3573

@asztal - feel free to submit a similar PR if you like and I can close mine.  Just wanted to get the ball rolling on getting this fix in `15.x` which is where my team needs it.

**Comment by IvanGoncharov** [comment]
@asztal Any news on CLA on your side.
If you still experiencing technical issues with the CLA console I can connect to people at the Linux foundation that can hopefully solve it.

**Comment by chrskrchr** [comment]
@asztal - any word on the CLA?

**Comment by hobby203** [comment]
Hi @chrskrchr

Sorry about the delays, @asztal has since left and this one's fallen through the cracks a little. We're still chasing CLA approval, it's still stuck with the CLA approval but currently trying to find out if a ticket has been submitted

**Comment by saihaj** [comment]
@hobby203 can you try to rebase this and force push so the bot is happy?

**Comment by hobby203** [comment]
Hi @saihaj,

I'm not sure if I can. This pr has been created from a fork under the developer's personal account, I don't think I'm able to do that. I could fork his fork and create a new PR, would that affect CLA approval? If it would I can try get in contact with @asztal and ask him to help out but would prefer not to bother him if possible.

Best,
Ed

**Comment by asztal** [comment]
Hopefully I'll have time to rebase the branch tonight. Thanks for getting the CLA sorted @hobby203 🙏

**Comment by hobby203** [comment]
Hi all,

Looks as though we're having issues with the CLA still. The EasyCLA tool won't approve it as @asztal doesn't have any commits on this project, but we know this to be the case as the fix is coming from their own fork and not from within the project. Is this something that's happened before? Not sure how to proceed.

**Comment by saihaj** [comment]
@hobby203 do you want to create PR with these changes and see if your CLA passes?

**Comment by hobby203** [comment]
@saihaj from inside this repo? or from my own fork, I checked but I can't create a new pr using these exact changes but i can try recreating the fix.

**Comment by hobby203** [comment]
@chrskrchr - if everyone is happy with that I think it will be the best course of action. I'm not sure why we're having these problems but we are struggling to make progress on our end with this issue, so if it can be moved forward by taking another approach i'm more than happy to let that happen (:

**Comment by chrskrchr** [comment]
> if everyone is happy with that I think it will be the best course of action.

No problem!  I'll go ahead and submit that new PR (giving credit to @asztal for the actual fix) and let the maintainers decide if they're OK with moving forward that way.

**Comment by chrskrchr** [comment]
A clone of this PR has been submitted as #3651.

@saihaj @IvanGoncharov - can you please look at [@hobby203's comment above](https://github.com/graphql/graphql-js/pull/3529#issuecomment-1158934485) that gives their blessing for this PR to be superseded by another PR to in order to move past the CLA issues?

**Comment by saihaj** [comment]
Sadly I can't be of much help with CLA bot 🥲 if the original author is fine having the PR from someone else I am fine moving forward with it.

**Comment by asztal** [comment]
> Sadly I can't be of much help with CLA bot 🥲 if the original author is fine having the PR from someone else I am fine moving forward with it.

I'm fine with that, as long as the bug is fixed I'm happy. Apologies for the delays on my end.

**Comment by benjie** [comment]
@asztal The CLA error

"The commit (https://github.com/graphql/graphql-js/commit/bc7049bcb29caf6a88bf4b6b16f7294e3b2314ce). This user is missing the User's ID, preventing the EasyCLA check. [Consult GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve.For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4)."

normally indicates that there's an issue with your authorship details for the given commit. Please ensure that that commit is authored by the exact same email address that you have approved via the CLA. Here's some instructions on how to change the email address on that commit:

https://stackoverflow.com/a/3042512

**Comment by IvanGoncharov** [comment]
Fixed by #3706

**Pull request closed without merging**